### DISCUSSION TITLE
my reweighting issue

### DISCUSSION THREAD

**Discussion by wkitlasten** [usgs/pestpp#250] [General]
My current ies setup is failing to update any pars; every lambda (1-1000) and scale fac (default) combo looks identical.

I checked a few individual realizations using pst.parrep() and the input and output files are changing as expected. The changes are reflected in the obs ensemble, where individual obs have reasonable values and distributions (+/- 10m head, etc). The parameter ensemble looks good too, as far as I can tell. Is there a method to calc or save an empirical jco (pestpp-ies seem inpervious to setting pst.control_data.jcosave and I'm not seeing anything in pyemu... I think I might have asked this already, just can't seem to find the answer).

Could it be a weighting issue? I am reusing a previous setup. In the current setup I am assigning weight=0 for all obs with flag "tritium_measured" and "greater_". PSo my "balance_phi_groups" table is:

tritium_measured,0
head,1
flow,1
less_flw,1
greater_,0

I get a "...WARNING: no non-zero obs groups found for tag 'TRITIUM_MEASURED'" but it is intended, so I ignore it.

Any suggestions on where else to look would be greatly appreciated!

**Comment by jtwhite79** [comment]
Have you looked at the lambda ensembles?  I guess you compare between them and also compare them to the prior to see how much change and in what direction ies is trying to move parameters?  Are the lambdas ensemnles running into bounds? Is the prior ensemble "centered" between the par bounds?

Might also be worth looking at the weighted residual ensemble (you'll have to calculate this with the updated obs info and prior ibs ensemble) - the residuals will control the magnitude of the par changes. 

What lambda are being used?  That phi table sums to 3 so the initial phase are being multipled by 3...

I've not seen a real case where ies refuses to go downhill at all...but I've manufactured cases where the posterior is highly nonlinear/multimodal that cause no phi decreases because each realization needs to move in a different direction and the global ensemble gradient doesn't represent those directioons. Maybe a test with a much tighter prior (artificially collapsed around the initial par values) would reveal this issue?

**Reply by wkitlasten** [reply]
Lots to dig through there, thanks.

All the lambda ensembles are the same as the prior (within 1e-17). Phi-based initial lambda was 1 (default lambda mult and scale). When that didn't do anything, I tried ies_initial_lambda=-1000 and -0.001 to cover a huge range... they all came out the same. Sorry, I'm not sure about: "That phi table sums to 3 so the initial phase are being multipled by 3"

Some reals have some pars at or near bounds. The distributions are somewhat centered but not perfectly. From previous experience these don't seem too bad (log10 of some K multipliers)??:

![some_k](https://user-images.githubusercontent.com/14982084/235545506-9ee90e54-778c-49dd-b640-97174170966a.png)

I'm not sure how to calc the weighted residual ensemble "with the updated obs info" since no pars are changing so the saved obs ens doesn't change. Is there a way to intercept simulation results returned from the workers? Here are some weighted residuals between the simulated obs and the obs+noise calculated using (?): `obs.loc[:, 'weight'].mul(oe0-ne)`. First half are flow obs, last half are head obs:

![wtres](https://user-images.githubusercontent.com/14982084/235555751-61aa7bcd-5bfd-4fae-bce4-82b99c27eb84.png)


No doubt the only thing special about this case is the mess I've made (like an untalented and more chaotic Jackson Pollock)! I am going to start over from scratch... I appreciate your patience and feedback while I dig through it.

**Reply by jtwhite79** [reply]
So the upgrade ensembles are numerically identical? I guess that explains why no phi reduction.  Some reasons for that a really really large lambda value or the SVD truncation is too aggressive - what values do you have from eigthresh and maxsing?  If you turn up the ies verbose level to 4, I think ies will write the upgrade component matrices and vectors out (this will the performance down...) but then you can have a look at these to see how many components are being used.  Also any localization here?

**Reply by wkitlasten** [reply]
Yes, numerically identical updgrade ensembles, even with small lambda (now tested 1e-7 to 1e4).

I think SVD is reasonable:

maxsing                             20000
eigthresh                           1e-06

No localization being used.

Verbose turned up to 4 and the journey continues.

**Reply by wkitlasten** [reply]
Not sure what I am looking with the new files; can't find descriptions in the user's guide. Looks like `upgrade_1` is the upgrade for each adjustable par for each real? No headers though, can I assume they are aligned with the par ensemble?

The values are tiny, obviously a problem! Does that suggest a bunk jco?
```
>>> ug=pd.read_csv('thread_0.count_ 0.iter_1.upgrade_1.dat',delim_whitespace=True)
>>> ug.max().max()
2.47264e-29
>>> ug.min().min()
-2.24079e-29
```

What are the `par_diff` and `ivec` files?

**Reply by jtwhite79** [reply]
Those files are raw ascii dumps of the numerical quantities from the solver - only for debugging so no real documentation...

The `upgrade` files are the parameter ensemble "upgrades" that get added to the parameter ensemble, element by element. 

The `ivec` files are scale and transformed singular values of the centered sim vals (which are in the `obs_diff` files).  Can you check those two file to look for strangeness?

**Reply by wkitlasten** [reply]
```
>>> iv.max().max()
5.12194e-46
>>> iv.min().min()
0.0
```
![ivec](https://user-images.githubusercontent.com/14982084/235807117-b0ba2f6a-a186-44e5-9a03-1cfad678c585.png)


I don't see any `obs_diff` files. Did you mean `par_diff` or is something missing?
```
>>> pdiff.min().min()
-10.5752
>>> pdiff.max().max()
10.408
```

Cleaning up a few weird looking par distributions. Maybe those are messing everything up?

**Reply by jtwhite79** [reply]
well the singular values show why the upgrades are numerically zero...whats that `obs_diff` matrix look like - those are the centered simulated values (from running the prior par ensemble...) and that matrix is what is factorized with SVD...

**Reply by wkitlasten** [reply]
When is the `obs_diff` file produced? Does it require calcs for all lambda mult and scale factors (still waiting on the last of those)?

![image](https://user-images.githubusercontent.com/14982084/236063641-f701db76-edfd-4f94-9c27-b8420bd8c70f.png)

**Reply by jtwhite79** [reply]
Should be just before ivec...there should be an obs_resid also with the same naming scheme also.

**Reply by wkitlasten** [reply]
The only files I am seeing with that naming convention are listed above... no obs_resid or obs_diff.

**Reply by jtwhite79** [reply]
thats weird.  There is a try-catch in that matrix writer routine so i wonder if there is something bunk (nan, inf, etc) in the obs matrix and the writer is just skipping them.  How big is this problem?  I might be interested in trying to debug it and see what is happening...
